src/opm/input/eclipse/Schedule/UDQ/UDQState.cpp

```diff
@@ -17,201 +17,355 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-
-#include <stdexcept>
-
 #include <opm/input/eclipse/Schedule/UDQ/UDQState.hpp>
+
 #include <opm/io/eclipse/rst/state.hpp>
 
-namespace Opm {
+#include <cstddef>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <unordered_map>
+
+#include <fmt/format.h>
 
 namespace {
 
-bool is_udq(const std::string& key) {
-    if (key.size() < 2)
-        return false;
+template <typename K, typename V>
+using UMap = std::unordered_map<K, V>;
 
-    if (key[1] != 'U')
-        return false;
+template <typename V>
+using SMap = UMap<std::string, V>;
+
+template <typename T>
+using S2Map = SMap<SMap<T>>;
+
+template <typename K, typename V>
+using SKMap = SMap<UMap<K, V>>;
 
-    return true;
+template <typename K, typename V>
+using S2KMap = S2Map<UMap<K, V>>;
+
+bool is_udq(const std::string& key)
+{
+    return (key.size() >= std::string::size_type{2})
+        && (key[1] == 'U');
 }
 
-bool has_var(const std::unordered_map<std::string, std::unordered_map<std::string, double>>& values,
-             const std::string& wgname,
-             const std::string udq_key) {
+bool has_var(const S2Map<double>& values,
+             const std::string&   wgname,
+             const std::string&   udq_key)
+{
     auto res_iter = values.find(udq_key);
-    if (res_iter == values.end())
-        return false;
+    return (res_iter != values.end())
+        && (res_iter->second.count(wgname) > 0);
+}
+
+void add_undefined_results(const Opm::UDQScalar& result,
+                           SMap<double>&         values)
+{
+    values.erase(result.wgname());
+}
+
+void add_undefined_results(const Opm::UDQScalar&       result,
+                           SKMap<std::size_t, double>& values)
+{
+    auto wellPos = values.find(result.wgname());
+    if (wellPos == values.end()) {
+        // No results for this well.  Nothing to do.
+        return;
+    }
+
+    wellPos->second.erase(result.number());
+}
+
+void add_defined_results(const Opm::UDQScalar& result,
+                         SMap<double>&         values)
+{
+    values.insert_or_assign(result.wgname(), result.get());
+}
+
+void add_defined_results(const Opm::UDQScalar&       result,
+                         SKMap<std::size_t, double>& values)
+{
+    auto wellPos = values.find(result.wgname());
+    if (wellPos == values.end()) {
+        values.insert_or_assign(result.wgname(), std::unordered_map<std::size_t, double>
+                                {{ result.number(), result.get() }});
+
+        return;
```

Comment:
Why not use
```
auto [iter, success] =  values.insert(...);
if(success) return;
....
```

Reply:
> Why not use
>
>```C++
> auto [iter, success] =  values.insert(...);
> if(success) return;
>```

I considered that, well with `emplace()` instead of `insert()`, but ultimately decided against it because it leads to unconditionally constructing the `unordered_map<std::size_t, double>` value object and destroying that object if the `wgname()` already exists in `values`.  E.g., we'd get the following code
```C++
auto [wellPos, inserted] =
    values.emplace(result.wgname(),
                   std::unordered_map<std::size_t, double> {
                       { result.number(), result.get() }
                   });

if (! inserted) {
    wellPos->second.insert_or_assign(result.number(), result.get());
}
```
and to be honest I don't think that's any better.